core/src/main/java/io/substrait/relation/RelProtoConverter.java

```diff
@@ -522,6 +530,111 @@ public Rel visit(NamedUpdate update, EmptyVisitationContext context) throws Runt
+    }
+    if (exchange.getTargets().isPresent()) {
+      builder.addAllTargets(
+          exchange.getTargets().get().stream().map(this::toProto).collect(Collectors.toList()));
+    }
+    return Rel.newBuilder().setExchange(builder).build();
+  }
+
+  @Override
+  public Rel visit(MultiBucketExchange exchange, EmptyVisitationContext context)
+      throws RuntimeException {
+    ExchangeRel.Builder builder =
+        ExchangeRel.newBuilder()
+            .setMultiTarget(
+                ExchangeRel.MultiBucketExpression.newBuilder()
+                    .setExpression(toProto(exchange.getExpression()))
+                    .setConstrainedToCount(exchange.getConstrainedToCount())
+                    .build());
+    if (exchange.getPartitionCount().isPresent()) {
+      builder.setPartitionCount(exchange.getPartitionCount().get());
+    }
+    if (exchange.getTargets().isPresent()) {
+      builder.addAllTargets(
+          exchange.getTargets().get().stream().map(this::toProto).collect(Collectors.toList()));
+    }
+    return Rel.newBuilder().setExchange(builder).build();
+  }
+
+  @Override
+  public Rel visit(RoundRobinExchange exchange, EmptyVisitationContext context)
+      throws RuntimeException {
+    ExchangeRel.Builder builder =
+        ExchangeRel.newBuilder()
+            .setRoundRobin(
+                ExchangeRel.RoundRobin.newBuilder().setExact(exchange.getExact()).build());
+    if (exchange.getPartitionCount().isPresent()) {
+      builder.setPartitionCount(exchange.getPartitionCount().get());
+    }
+    if (exchange.getTargets().isPresent()) {
+      builder.addAllTargets(
+          exchange.getTargets().get().stream().map(this::toProto).collect(Collectors.toList()));
+    }
+    return Rel.newBuilder().setExchange(builder).build();
+  }
+
+  @Override
+  public Rel visit(BroadcastExchange exchange, EmptyVisitationContext context)
+      throws RuntimeException {
+    ExchangeRel.Builder builder =
+        ExchangeRel.newBuilder().setBroadcast(ExchangeRel.Broadcast.newBuilder().build());
+    if (exchange.getPartitionCount().isPresent()) {
+      builder.setPartitionCount(exchange.getPartitionCount().get());
+    }
+    if (exchange.getTargets().isPresent()) {
+      builder.addAllTargets(
+          exchange.getTargets().get().stream().map(this::toProto).collect(Collectors.toList()));
+    }
+    return Rel.newBuilder().setExchange(builder).build();
+  }
+
+  private ExchangeRel.ExchangeTarget toProto(AbstractExchangeRel.ExchangeTarget target) {
+    ExchangeRel.ExchangeTarget.Builder builder =
+        ExchangeRel.ExchangeTarget.newBuilder().addAllPartitionId(target.getPartitionIds());
+    if (target.getType() instanceof TargetType.Uri) {
+      builder.setUri(((TargetType.Uri) target.getType()).getUri());
+    }
+    if (target.getType() instanceof TargetType.Extended) {
+      builder.setExtended(((TargetType.Extended) target.getType()).getExtended());
+    }
```

Comment:
given that `TargetType.Uri` and `TargetType.Extended` are classes `target.getType()` can not be both so I would write this with an `else if`:
```suggestion
    if (target.getType() instanceof TargetType.Uri) {
      builder.setUri(((TargetType.Uri) target.getType()).getUri());
    } else if (target.getType() instanceof TargetType.Extended) {
      builder.setExtended(((TargetType.Extended) target.getType()).getExtended());
    }
```